### PR TITLE
Cosmo read/write ECSV

### PR DESCRIPTION
**Pull request by nstarman**
### Description

A Cosmology can be read from / written to an ECSV file. This keeps all units and metadata and will work when Cosmology is enhanced to be able to be array-valued.

~Requires #12213 . Will be rebased when that it is merged. Test failing because there is more than 1 changelog fragment.~
Merged.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by nstarman** [comment]
@nstarman @taldcroft
If the ``mypackage`` move is too large to review, I can split it off of this PR.

File: docs/whatsnew/5.0.rst [hunk 1]
```diff
@@ -20,15 +20,14 @@ files and converting between different Python objects.
...
-    >>> from astropy.cosmology import Planck18
-    >>> Planck18.write('<file name>.<format>')
+    >>> from astropy.cosmology import Cosmology, Planck18
+    >>> Planck18.write('<file name>.ecsv')
     >>>
-    >>> from astropy.cosmology import Cosmology
-    >>> cosmo = Cosmology.read('<file name>.<format>')
+    >>> cosmo = Cosmology.read('<file name>.ecsv')
     >>> cosmo == Planck18
     True
```
Comment (taldcroft) [inline docs/whatsnew/5.0.rst:33]: I think the examples below are a bit long for the What's New. The idea is not to comprehensively show the capabilities but just highlight things enough for interested people to click through to the docs link. I think each What's New entry should fit comfortably on one laptop screen (and maybe half the people are reading on mobile?). So maybe just:
```
Transformations between ``Cosmology`` and `dict`, ``QTable``,  and``Model`` are
pre-registered, enabling for example::

    >>> from astropy.cosmology import Cosmology, Planck18
    >>> cm = Planck18.to_format("mapping")
    >>> cm
    {'cosmology': <class 'astropy.cosmology.flrw.FlatLambdaCDM'>,
     'name': 'Planck18',
     'H0': <Quantity 67.66 km / (Mpc s)>,
     'Om0': 0.30966,
     ...

For details see :ref:`whatever`.

**Review by taldcroft — COMMENTED** [review]
Just a few docs comments for now, will try to get to the rest soon.

File: docs/whatsnew/5.0.rst (hunk 1 above)
Comment (nstarman) [reply docs/whatsnew/5.0.rst:33]: SGTM. I would like to show Table and Model. I think those are the compelling examples.

File: docs/whatsnew/5.0.rst [hunk 2]
```diff
@@ -20,15 +20,14 @@ files and converting between different Python objects.
...
 .. doctest-skip::
 
-    >>> from astropy.cosmology import Planck18
-    >>> Planck18.write('<file name>.<format>')
+    >>> from astropy.cosmology import Cosmology, Planck18
+    >>> Planck18.write('<file name>.ecsv')
     >>>
-    >>> from astropy.cosmology import Cosmology
-    >>> cosmo = Cosmology.read('<file name>.<format>')
+    >>> cosmo = Cosmology.read('<file name>.ecsv')
```
[the remark this answers is not in the document]
Comment (nstarman) [reply docs/whatsnew/5.0.rst]: I didn't think whatsnew was skipped in tests. This makes examples much better.

File: docs/whatsnew/5.0.rst [hunk 3]
```diff
@@ -20,15 +20,14 @@ files and converting between different Python objects.
 The methods use Astropy's Unified I/O registry so custom formats can be
 registered. Details are discussed in an addition to the docs.
```
[the remark this answers is not in the document]
Comment (nstarman) [reply docs/whatsnew/5.0.rst]: Done.

File: docs/whatsnew/5.0.rst [hunk 4]
```diff
@@ -20,15 +20,14 @@ files and converting between different Python objects.
...
 
-Currently no file formats are registered, but the syntax is as follows:
+Currently the only registered read/write format is "ascii.ecsv", like for Table.
```
[the remark this answers is not in the document]
Comment (nstarman) [reply docs/whatsnew/5.0.rst]: Thanks!
Implemented.

**Review by taldcroft — APPROVED** [review]
Looks good!

**Comment by nstarman** [comment]
Thanks @taldcroft for pushing to get this in for v5.0!

**Pull request merged**